src/analyses/goto_check.cpp

```diff
@@ -351,6 +365,53 @@ void goto_checkt::mod_by_zero_check(
     guard);
 }
 
+/// Check if the frontend has flagged a given expression for range checking.
+/// \param e: The expression that is to be checked.
+/// \return: True or false based on whether the flag was set or not.
+static bool range_check_flag_set(const exprt &e)
+{
+  return e.get_bool("range_check");
+}
```

Comment:
I would introduce a separate expression for that, as opposed to attaching a non-comment attribute to arbitrary expressions.

Reply:
I am not sure I understand how to do it. There's a variety of different expressions that can use a range check, and to create a new expression for them would mean that we would need to subclass them right? The problem is that these expressions are different classes, with sometimes different parents, which would mean that an inheritance based solution would be messy.

Would it be better if we just made both the `range_check` and the `overflow_check` comments for now, so that they don't affect structural equality checks on the ireps at least? What do you think @kroening ?

Reply:
As I see it there are two issues here:

A. What to do with run-time checks that are inserted by the Ada front-end?
B. If and how we transfer information about ranged types from the Ada front-end to CPROVER.

For A. I think the answer is relatively clear:
A.1. gnat2goto should produce calls to internal functions (__CPROVER_Ada_range_check or something ) for all of the Ada run-time checks.
A.2. These should pass through the JSON Symtab code and wind up in the goto-program.
A.3. We should provide a "default" implementation of these so that things just work in the expected way.
A.4. We should also provide an instrumentation pass that recognises these specific functions calls and converts them into assert, assert then assume, assume, assert then throw if false or just throw if false.

Why : the actual semantics are throw if false but depending on what you are trying to do (find runtime errors, prove things about the "correct" behaviour of the code, etc.) and what run-time you are using (is there anything other than the default catch-and-terminate?) the other behaviours are more desireable.

I'm a bit more undecided on B.  In some ways it would be nice to have the Ada typing carried over into CPROVER.  However I'm not sure if the benefits are worth the hassle of doing it just yet.  To me the key question is what we would just that information for.  GNAT should have already inserted the right checks to capture overflow and underflow.  So what would we use it for?  Better nondets?  That said, just because we can't come up with anything at this stage doesn't mean we should write this off for ever and we may have to come back and revisit this.

Reply:
@martin-cs I'd personally prefer just generating the right thing from gnat2goto controllable via a switch rather than producing a bunch of ada specific intrinsics that need to be supported in cbmc.

As to B: I'd also prefer Ada typing to stay within Ada-land for the most part, especially as some Ada types are dynamic (which'd be a bit weird from the cbmc perspective and needs to be implemented on the gnat2goto side anyway). I'd suggest we generate globals that have the right range values and things like that for static types, and just do what the ada runtime would have to do to support dynamic types, i.e. pass them on as additional parameters.

Reply:
@hannes-steffenhagen-diffblue 
A. There isn't a one right thing to do; it depends on the analysis you are doing and thus should not be part of the front-end IMHO.  I would prefer to keep the compile process as simple and as easily integrated with various build systems as possible; additional flags will not help that.

B. CPROVER already supports RTTI for C++ (and I suspect Java) so tagged types should be handled like that.  I guess my main concern is with ranged types and Ada 2012 type predicates.  The current "base type + range check" solution is probably fine for now but it's something we may need to keep in mind (and one of the reasons I want the range checks clearly marked so that the abstract interpreter can pick up information from them if needs be).